### PR TITLE
Add L3 forwarding to cndpfwd example

### PR DESCRIPTION
**Pull request by stvml**
This commit enhances the cndpfwd example application by adding a new `l3-fwd` mode. In this mode, incoming packets are forwarded based on a FIB defined in `fwd.jsonc`, in the `options:l3fwd-fib-rules` section. Each FIB entry consists of an IP/prefix to match, a destination MAC for the next hop, and a TX port to use.

Targets #231.

### CODE REVIEW

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:30]: Not a big fan of having these constant values in the source code. One option is to add a new key value pair array to the .jsonc file in the **options** section. This was done in the examples/cnet-graph jsonc file for adding socket connections.

I was thinking we can add a new array in the options section for this rules.

// fib-rules define the IP address, MAC address and port index for the FIB table entries
//    Three parts to each entry in the array. IP address with mask value/depth, MAC address and port index.
"fib-rules": [
    "198.18.0.0/24,02:00:01:02:03:04,1",
    "198.18.1.0/24,06:00:01:02:03:04,0",
]

The above is just an example and we can format the strings as you see fit. The example/cnet-graph has an example of how to access the options:graph:0-chanl strings.

Then we can remove hardcoded values from this file.

**Comment by KeithWiles** [comment]
Excellent work, thanks.

**Comment by KeithWiles** [comment]
It is a good idea to have the Git hook installed to clang-format the code as it is checked into the repo. This is what causes the check failures above.

The file cndp/.githooks/pre-commit needs to be copied to the correct location. Look in the CONTRIBUTING.md file around line 22 for more details.

Thank you for the patch it is looking good.

File: examples/cndpfwd/l3-fwd.c
Comment (stvml) [reply examples/cndpfwd/l3-fwd.c:30]: We did this because it was a little easier to implement (and after discussion with @jeffreybshaw who pointed me to similar behavior in https://github.com/DPDK/dpdk/blob/main/examples/l3fwd/main.c#L204).

Totally agree that parsing the rules from config would be cleaner, though. Want us to do that on this PR, or open an enhancement issue once this one is merged?

**Comment by stvml** [comment]
> It is a good idea to have the Git hook installed to clang-format the code as it is checked into the repo. This is what causes the check failures above.

Yep, read `CONTRIBUTING.md` but got excited and forgot to install hooks/actions when setting up our fork. :) Expect a `clang-format`ted update to this PR, probably sometime on Monday. Cheers!

File: examples/cndpfwd/l3-fwd.c
Comment (stvml) [inline examples/cndpfwd/l3-fwd.c:57]: While I'm here, also wanted to ask a maintainer's opinion on this call. `cne_fib_lookup_bulk` appears to be designed to handle arrays of `N` lookups, where in this case `N` is always 1. We call this function in the `for` loop over the array of received packets.

We didn't have time to look into this, but would it be more idiomatic if we were to set up those arrays in that RX loop, make a single call to `cne_fib_lookup_bulk`, then rewrite and TX the packets in a loop, as opposed to the current approach where we RX->lookup->rewrite->TX in a loop?

File: examples/cndpfwd/main.c
Comment (stvml) [inline examples/cndpfwd/main.c:182]: Just noticed that we lost our logic here that should check whether the `jcfg_lport_by_index` call succeeded, which means this line could try to dereference NULL if the user specifies a nonexistent TX port index in their FIB definition.

The regular `fwd` test echoes packets back out of the RX port if there's no TX port matching the low byte of the MAC, but in our case, since the TX port indices are defined ahead of time, I'm inclined to drop them...

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [reply examples/cndpfwd/l3-fwd.c:30]: Have a look and see if you think opening a new PR or enhancement is reasonable, I would prefer it fixed in this PR. I can understand if you want to leave it this way and create a new PR or enhancement. If you are going to create a new PR please create an enhancement for the change.

File: examples/cndpfwd/main.c
Comment (KeithWiles) [reply examples/cndpfwd/main.c:182]: If we drop the packets with invalid TX port then we can not tell where the packet is dropped. Sending the packet back out the RX port seemed like a reasonable solution as we can see the packets getting looped back to the RX port. I can see both ways could work, let's drop them for now. Maybe some type of indicator to tell the user what is happening as it is an error.

When we add the jsonc **fib-rules**, we need to check the TX port exists and if not error out the rule.

The check for **dst** being null should be done, with a message could work. It could a be a huge number of messages, but I think it is reasonable.

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [reply examples/cndpfwd/l3-fwd.c:57]: I believe it would be a good change. It just means the arrays for the FIB lookup would need to be created dynamically using a dynamic array based on n_pkts or maximum statically created arrays. Using calloc or malloc here would be a performance hit.

Comment (stvml) [reply examples/cndpfwd/l3-fwd.c:89]: Will create an enhancement issue for making this change once this PR is merged.

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:62]: We can use the macro CNE_ERR_RET("Invalid TX port index value: %s\n", entry[FIB_RULES_PORT]); in place of return -1;

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:68]: We also have a MAC address convert routines?

```
struct ether_addr eaddr;
struct ether_addr *addr = cne_ether_aton(mac_addr, &eaddr);
if (addr == NULL)
    CNE_ERR_RET("Ethernet address is invalid (%s)\n", mac_addr);
```

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:72]: With the above this line is not required.

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:11]: Not really a JSON index. Maybe /**< Index to IP address in parsed string array */, ... or something like it.

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:78]: Use CNE_ERR_RET() instead of return -1;

File: examples/cndpfwd/main.c
Comment (KeithWiles) [inline examples/cndpfwd/main.c:169]: Minor nit. On line 192 we use 'int i = 0', but here we use i = 0 with a global int i shadowing i. I would suggest to remove the global int i and change this one to int i = 0 to be consistent.

File: examples/cndpfwd/parse-args.c
Comment (KeithWiles) [inline examples/cndpfwd/parse-args.c:81]: We normally use calloc() to make sure the memory is zero. May not matter much here, but just to be safe.

File: examples/cndpfwd/parse-args.c
Comment (KeithWiles) [inline examples/cndpfwd/parse-args.c:82]: Move this line to after line 80 and remove the = 0; as it is not required. The forloop will set the value to zero.

Remove the {} from around the for loop as they are not required here and match the style.

File: examples/cndpfwd/l3-fwd.c
Comment (KeithWiles) [inline examples/cndpfwd/l3-fwd.c:47]: We do not use {} around single like for, while, if, else, ... lines. Please remove if this is needed.

**Review by KeithWiles — CHANGES_REQUESTED** [review]
Thank you for adding the JSON config support, but I did have a few minor changes and suggestions for error handling.

**Comment by KeithWiles** [comment]
in the Smoke test above, we have a compile problem. Most likely a different compiler version.

```
../examples/cndpfwd/l3-fwd.c: In function ‘l3fwd_fib_init’:
../examples/cndpfwd/l3-fwd.c:113:17: error: invalid use of undefined type ‘struct cne_fib’
  113 |         free(fib->fib_rules[i]);
      |                 ^~
../examples/cndpfwd/l3-fwd.c:114:13: error: invalid use of undefined type ‘struct cne_fib’
  114 |     free(fib->fib_rules);
      |             ^~
```

File: examples/cndpfwd/parse-args.c
Comment (KeithWiles) [inline examples/cndpfwd/parse-args.c:82]: Need to check to make sure f->fib_rules is not NULL from calloc().

**Review by KeithWiles — CHANGES_REQUESTED** [review]
Thank you, one more minor issue.

**Review by KeithWiles — APPROVED** [review]
LGTM thank you

**Comment by maryamtahhan** [comment]
@stvml @KeithWiles 
Hey guys
I'm seeing some pretty weird behaviour/performance with the scenario (cndpfwd in l3fwd mode between a physical port and veth):

                      +-------------cndpfwd-l3fwd---------+
traffic gen --> lport 1 (physical NIC) --> lport 2 (veth) ---------------+
                                                                                                                      | loopback
                                     traffic gen <-- lport 1 (physical NIC)  < --------+

Did ye notice any unusual performance drops in the phy to phy testing (I don't have enough ports to replicate phy to phy atm)?

**Comment by stvml** [comment]
Honestly, the extent of our testing was veth to veth, and inside a VM... and we didn't have time to bring up a TRex instance and see what happens under realistic amounts of traffic. We just sent out single handmade Scapy packets to confirm that the FIB lookup and rewrite logic was doing what it was supposed to. 

Sorry we don't have anything useful to say on performance, but I'm thrilled that someone's tried putting this under load--what behavior are you seeing?

**Pull request merged**
